lib/Conversion/AIEVecToLLVM/AIEVecToLLVM.cpp

```diff
@@ -354,6 +354,110 @@ class MulOpConversion : public mlir::ConvertOpToLLVMPattern<aievec::MulOp> {
   }
 };
 
+class MulElemOpConversion
+    : public mlir::ConvertOpToLLVMPattern<aievec::MulElemOp> {
+public:
+  using ConvertOpToLLVMPattern<aievec::MulElemOp>::ConvertOpToLLVMPattern;
+
+  struct DecodedMulElemOp {
+    enum class Kind {
+      // DtIn0_DtIn1_DtRes_CxMxKxN
+      I8_I8_I32_32x1x2x1,
+      I16_I16_I32_32x1x1x1,
+      BF16_BF16_FP32_16x1x2x1,
+      UNSUPPORTED
+      // TODO: I16_I16_I64_16x1x2x1
+      // TODO: I32 and FP32 mul_elem are emulated
+    };
+
+    Kind kind;
+    int conf;
+  };
+
+  static int aiev2_mul_mac_compute_control(int sgn_x, int sgn_y, int amode,
+                                           int bmode, int variant, int zero_acc,
+                                           int shift16, int sub0, int sub1,
+                                           int sub2, int sub_mask) {
+    return ((unsigned)sub_mask << 16) | ((unsigned)shift16 << 10) |
+           ((unsigned)sub0 << 11) | ((unsigned)sub1 << 12) |
+           ((unsigned)sub2 << 13) | ((unsigned)amode << 1) |
+           ((unsigned)bmode << 3) | ((unsigned)variant << 5) |
+           (((unsigned)sgn_x << 9) | ((unsigned)sgn_y << 8)) |
+           ((unsigned)zero_acc << 0);
+  }
+
+  static DecodedMulElemOp decodeMulElemOp(OpAdaptor op) {
+    auto lhs = op.getLhs();
+    auto lhsVecTy = cast<VectorType>(lhs.getType());
+    auto lhsScaTy = lhsVecTy.getElementType();
+    unsigned lhsBitWidth = lhsScaTy.getIntOrFloatBitWidth();
+
+    // Integer types
+    if (lhsScaTy.isa<IntegerType>()) {
+      if (lhsBitWidth == 8) {
+        return {DecodedMulElemOp::Kind::I8_I8_I32_32x1x2x1,
+                aiev2_mul_mac_compute_control(1, 1, 0, 1, 1, 0, 0, 0, 0, 0, 0)};
+      } else if (lhsBitWidth == 16) {
+        return {DecodedMulElemOp::Kind::I16_I16_I32_32x1x1x1,
+                aiev2_mul_mac_compute_control(1, 1, 0, 3, 1, 0, 0, 0, 0, 0, 0)};
+      }
+    } else {
+      // Float types
+      if (lhsBitWidth == 16) {
+        return {DecodedMulElemOp::Kind::BF16_BF16_FP32_16x1x2x1,
+                aiev2_mul_mac_compute_control(0, 0, 2, 3, 1, 0, 0, 0, 0, 0, 0)};
+      }
+    }
+
+    return {DecodedMulElemOp::Kind::UNSUPPORTED, -1};
+  }
+
+  LogicalResult
+  matchAndRewrite(aievec::MulElemOp op, OpAdaptor adaptor,
```

Comment:
@jamestcl-amd , @jsetoain : This function is not doing any checking on the number of vector lanes or alternatively the total number of bits of the operands or result. Looking at https://github.com/Xilinx/mlir-aie/blob/11822b078459a6a3ead7d6d3e06ff20c1cb74172/lib/Dialect/AIEVec/IR/AIEVecOps.cpp#L903 I see that this checks consistency, but not the actual number of lanes. So, which piece of are we relying on checking that the number of lanes is correct?